[1.21] Overhaul tps command to be more colorful and organized

ApexModder opened the pull request:
The `/neoforge tps <dim>` command output has been overhauled to be much more colorful, making it easier to read and understand at a glance.

Both the TPS and Tick Time components now have a “hover event” to better explain the values. Additionally, the mean TPS fades from green to red, making it quicker and easier to identify laggy dimensions.

<details>
<summary>Before</summary>

![image](https://github.com/user-attachments/assets/d5fa4c9b-5d3d-4833-8ba1-58a958caee54)
</details>

<details>
<summary>After</summary>

![java_V1wncOEK0l](https://github.com/user-attachments/assets/e9b0c928-2cce-4757-a0eb-0fba31decf42)
</details>

HenryLoenwind commented:
Please don't drop the "ms" unit.

Please revert those commas back to periods or use lowercase after them. ", Mean" is just mean ;)

Also, I would suggest using `ResourceLocation.toLanguageKey()` instead of the raw RL (and putting the raw RL into the hover text). That way, it will translate automatically if we ever manage to put those language keys in for vanilla dimensions.

The "Dimension:" prefix seems unnecessary to me. I would drop it to avoid line breaking.

ApexModder commented:
Added back the `ms` unit; I didn’t actually intend on removing it.

Dropped the dimension prefix and updated to auto-translate if a matching translation key exists (defaults to the raw registry name if missing): `dimension.<namespace>.<path>`

I have included translations for the vanilla dimensions, but modders will have to include their own.

The raw dimension registry name has also been added as a hover tooltip when hovering over the dimension component.

Below is an example of these changes. Note how the Nether is `minecraft:the_nether`; this is because no `dimension.minecraft.the_nether` translation exists.

<details>

<summary>Example</summary>

![image](https://github.com/user-attachments/assets/d8124acd-23e9-41eb-912b-cd0d36f365e7)

</details>

ApexModder commented:
Locked behind #1428, Will need updating if that PR goes through

ApexModder committed:
Reimplement PR changes

ApexModder committed:
Update to make use of changes from #1428

ApexModder committed:
Apply Formatting

sciwhiz12 commented on src/main/java/net/neoforged/neoforge/server/command/TPSCommand.java:
```diff
@@ -5,53 +5,53 @@
     static ArgumentBuilder<CommandSourceStack, ?> register() {
         return Commands.literal("tps")
-                .requires(cs -> cs.hasPermission(0)) //permission
+                .requires(cs -> cs.hasPermission(Commands.LEVEL_ALL)) //permission
```
This can be removed, since the lowest natural (and default) permission level is 0 (or `LEVEL_ALL`) anyway.

sciwhiz12 commented on src/main/java/net/neoforged/neoforge/server/command/TPSCommand.java:
```diff
@@ -60,4 +60,65 @@ private static long mean(long[] values) {
+            tickRateManager = dimension.tickRateManager();
+        }
+
+        var tickTime = mean(times) * 1.0E-6D;
```
This line can be improved by first using the `Stats.meanOf` method from Guava, replacing the self-made `mean` method, and replacing the `* 1.0E-6D` part with `/ TimeUtil.NANOSECONDS_PER_MILLISECOND`.

The second change removes the magic constant to use a named constant instead, which makes the code more understandable, and in my previous testing avoided some possible floating-point rounding errors.

sciwhiz12 commented on src/main/java/net/neoforged/neoforge/server/command/TPSCommand.java:
```diff
@@ -60,4 +60,65 @@ private static long mean(long[] values) {
+
+            prefix = Component.translatable("commands.neoforge.tps.summary.dimension", createDimensionComponent(dimension));
+            times = dimensionTimes == null ? UNLOADED : dimensionTimes;
+            tickRateManager = dimension.tickRateManager();
```
I don't think this matters, since that method just returns the server's tick rate manager. I leave it up to you if you want to remove this line and just use the tick rate manager obtained directly from the server instance.

ApexModder replied:
You are correct. It does retrieve the servers tick rate manager, but what if some mod allowed the tick rate manager to be per dimension rather than server wide, this would already support that.
It also just feels more correct imo to pull off of the dimension instance if one exists rather than the server.

sciwhiz12 commented on src/main/java/net/neoforged/neoforge/server/command/TPSCommand.java:
```diff
@@ -60,4 +60,65 @@ private static long mean(long[] values) {
+                .withColor(ChatFormatting.GREEN)
+                .withHoverEvent(new HoverEvent(
+                        HoverEvent.Action.SHOW_TEXT,
+                        Component.literal(regName.toString())
+                                .append(" (")
+                                .append(dimensionType.toString())
+                                .append(")"))));
```
I think this should be a translation key.

sciwhiz12 commented on src/main/resources/assets/neoforge/lang/en_us.json:
```diff
@@ -81,10 +81,13 @@
+  "commands.neoforge.tps.mean_tps": "Mean TPS: %s",
+  "commands.neoforge.tps.mean_tps.tooltip": "Closer to 20 is better",
+  "commands.neoforge.tps.mean_tick_time": "Mean tick time: %s ms",
+  "commands.neoforge.tps.mean_tick_time.tooltip": "Lower is better, Higher is worse",
   "commands.neoforge.mods.list": "Mod List: %1$s",
-  "commands.neoforge.tps.summary.basic": "Dim %1$s: Mean tick time: %2$s ms. Mean TPS: %3$s",
-  "commands.neoforge.tps.summary.named": "Dim %1$s (%2$s): Mean tick time: %3$s ms. Mean TPS: %4$s",
```
Why did you split up the translation keys for the summary lines instead of reworking the existing ones, and applying color to the (`Component`) arguments? 

This is a bit more difficult to visualize when localizing through Crowdin, since one has to know how each part fits together to make the whole, instead of having the whole line be laid out in the translation key.

ApexModder replied:
I needed to split up into smaller chunks in order to apply the hover tooltips correctly, its applied to the section not translation arguments.
<sub>Notice how the tooltip in the below gif is visible when hovering more than just the numbers</sub>

Unless you know of a way to apply the hover to correct section while keeping it as 1 translation.
I mean I could apply the hover to the translation arguments but that decreases the hover area significantly making it harder to find (you would only be able to hover over the numbers).

![java_cHjLKeT3GS](https://github.com/user-attachments/assets/79b9f9b8-73d2-428c-9c7a-ef3459263706)

sciwhiz12 commented on src/main/java/net/neoforged/neoforge/server/command/TPSCommand.java:
```diff
@@ -5,53 +5,53 @@
-        cs.sendSuccess(() -> Component.translatable("commands.neoforge.tps.summary.named", dim.getDescription(), reg.getKey(dim.dimensionType()).toString(), TIME_FORMATTER.format(worldTickTime), TIME_FORMATTER.format(worldTPS)), false);
-
-        return 1;
+    private static int sendTime(CommandContext<CommandSourceStack> context, ServerLevel dim) throws CommandSyntaxException {
```
The `dim` parameter should also be marked as `@Nullable`, since it's given `null` in the command (for the overall summary).

Matyrobbrt commented:
I think it's better if the colours are part of the translation and use vanilla's weird character instead of being hard-coded, as it would be more customizable, especially for people that have a hard time with the built in colours.

ApexModder committed:
Remove permission requirement

ApexModder committed:
Use Guava Stats.meanOf

Also replace magic number with TimeUtil constant

ApexModder committed:
Mark parameter as Nullable

Rename 'dim' to 'dimension' for better readability

ApexModder committed:
Convert dimension tooltip to being a translatable component

ApexModder commented:
> I think it's better if the colours are part of the translation and use vanilla's weird character instead of being hard-coded, as it would be more customizable, especially for people that have a hard time with the built in colours.

While this would work for the coloring, how would I go about applying the hover tooltips that way?
afaik it is not possible to apply hover effects other than code.
But I do agree having the colors being configurable in some way would be better (even if its just the translation defining the colors).

HenryLoenwind commented:
Just a side note: The json-array format probably can do hover, but it should not be used as it falls apart on Crowdin. The SGML-format could be extended to support hover and even parameterised hovers; I just never thought about this use case.

ApexModder committed:
Big refactoring to shorten overall output

While keeping all of the same information and stylization

Matyrobbrt commented:
> afaik it is not possible to apply hover effects other than code.

That's correct, that much can stay code, but the colouring should still be part of the lang entry.

XFactHD commented on src/main/java/net/neoforged/neoforge/server/command/TPSCommand.java:
```diff
@@ -5,59 +5,94 @@
+        }
 
-        return 1;
+        return component.withStyle(style -> style.withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, Component.translatable("commands.neoforge.tps.tooltip", tickRateManager.tickrate()))));
```
Did you make sure that this hover event on the entire component does not prevent the dimension name hover event from showing up?

ApexModder replied:
Yes, you can still hover the dimension name to view the registry id and dimension type.
_See the updated "after" gif_

XFactHD reviewed (changes requested):
One detail I'm not sure about, otherwise the code looks good to me

sciwhiz12 commented on src/main/java/net/neoforged/neoforge/server/command/TPSCommand.java:
```diff
@@ -5,59 +5,94 @@
+
+        if (dimension == null)
+            component = Component.translatable("commands.neoforge.tps.overall", tpsComponent, tickTimeComponent);
+        else {
```
For consistency between the two branches, please use braces here as well.

sciwhiz12 commented on src/main/resources/assets/neoforge/lang/en_us.json:
```diff
@@ -80,13 +80,10 @@
+  "commands.neoforge.tps.overall": "Overall: %s TPS (%s ms/tick)",
+  "commands.neoforge.tps.tooltip": "Mean TPS; higher is better. Target TPS: %s",
+  "commands.neoforge.tps.dimension": "%s: %s TPS (%s ms/tick)",
+  "commands.neoforge.tps.dimension.tooltip": "%s (%s)",
```
As I understand from the code above, this tooltip contains the location of the dimension, and the location of the dimension _type_ in the parentheses. I didn't understand that on first glance, and I fear it might become a point of confusion for other users. 

Should we try to clarify what the location in parentheses means (the dimension _type_), or do we assume the user knows what it means?

ApexModder replied:
Looking at vanillas own translations they do include some context with the translation key.
Potions being a good example `"potion.withDuration": "%s (%s)"`.

I could update the key from `commands.neoforge.tps.dimension.tooltip` to `commands.neoforge.tps.dimension.withType` or something similar?

This might also be better suited as an extension for https://github.com/neoforged/NeoForge/pull/1428.
To add dimension type translation lookups. [`dimension_type.<namespace>.<path>`]
would also add simple utility translation
`"dimension.withType": "%s (%s)"` -> `<dimension component> (<dimension type component>)`

sciwhiz12 replied:
Just to be clear, I wasn't talking more about the translation key, but more on how it's shown in game: `minecraft:overworld (minecraft:overworld)` (for example) seems a bit odd to someone using the command that doesn't know the distinction between the two.

(It's acceptable with the vanilla potion duration translation, because the first value is the potion name and the second is a duration; both can't be confused for the other.)

ApexModder replied:
Oh. Yeah the format might be confusing for a regular user. However since the concept of TPS and the details of this command are intended for more intermediate users, I think its reasonable to assume they know the distinction between the 2.

sciwhiz12 replied:
Okay then. If it becomes a reported problem, we can probably just insert "dimension type: " then to clarify.

Eric-01 replied:
Please do this. I would call myself a "power user" who has been playing Minecraft since beta 1.4, but this is the first time I've come across something like dimension types. Perhaps I'm just ignorant, which is very likely :)

Eric-01 replied:
Please do this if possible. I would call myself a "power user" who has been playing Minecraft since beta 1.4, but this is the first time I've come across something like dimension types. Perhaps I'm just ignorant, which is very likely :)

ApexModder replied:
Appended `Dimension Type:` to the tooltip text, to make this distinction more obvious.

![image](https://github.com/user-attachments/assets/650e04a2-522e-4772-a490-e68191374a1e)

sciwhiz12 reviewed (approved):
Two small comments remaining on my end. Otherwise, it looks fine.

sciwhiz12 reviewed (changes requested):
Whoops, clicked the wrong button.

ApexModder committed:
Add braces to if statement

sciwhiz12 reviewed (dismissed):
Looks okay to me, and it works. Waiting on @XFactHD to rereview.

ApexModder committed:
Append 'Dimension Type:' to dimension tooltip

XFactHD reviewed (approved):
LGTM :shipit:
